ttnn/cpp/ttnn/operations/ccl/erisc_datamover_builder.hpp

```diff
@@ -0,0 +1,229 @@
+
+    [[nodiscard]] std::vector<uint32_t> emit_compile_time_args() const;
+
+    [[nodiscard]] std::vector<uint32_t> emit_runtime_args() const;
+
+    void connect_to_downstream_edm(FabricEriscDatamoverBuilder const& downstream_edm);
+
+    void dump_to_log() const {
+        // TODO
+    }
+
+   private:
+   friend class EdmLineFabricOpInterface;
+    CoreCoord my_eth_core_logical;
+    size_t my_noc_x;
+    size_t my_noc_y;
+
+    FabricEriscDatamoverConfig config;
+
+    size_t my_chip_id;
+    size_t peer_chip_id;
+    size_t handshake_address;
+    size_t channel_buffer_size;
+
+    size_t sender_0_num_buffers;
+    size_t sender_1_num_buffers;
+    size_t receiver_num_buffers;
+
+    size_t local_sender_channel_0_buffer_address;
+    size_t local_sender_channel_0_connection_info_addr;
+    size_t local_sender_channel_1_buffer_address;
+    size_t local_sender_channel_1_connection_info_addr;
+    size_t local_receiver_channel_buffer_address;
+
+    size_t termination_signal_ptr;
+
+    // Semaphore IDs
+    // this is the receiver channel's local sem for flow controlling with downstream fabric sender
+    std::optional<size_t> receiver_channel_downstream_flow_control_semaphore_id;
+    size_t sender_channel_0_flow_control_semaphore_id;
+    size_t sender_channel_1_flow_control_semaphore_id;
+    size_t sender_channel_0_connection_semaphore_id;
+    size_t sender_channel_1_connection_semaphore_id;
+    size_t sender_channel_0_buffer_index_semaphore_id;
+    size_t sender_channel_1_buffer_index_semaphore_id;
+    size_t receiver_channel_local_buffer_index_addr;
+
+    std::optional<size_t> downstream_edm_noc_x;
+    std::optional<size_t> downstream_edm_noc_y;
+    std::optional<size_t> downstream_edm_buffer_base_address;
+    std::optional<size_t> downstream_edm_semaphore_address;
+    std::optional<size_t> downstream_edm_worker_registration_address;
+    std::optional<size_t> downstream_edm_worker_location_info_address;
+    std::optional<size_t> downstream_sender_channel_buffer_index_semaphore_id;
+};
+
+
+struct edm_termination_info_t {
+    uint32_t distance;
+    uint32_t edm_noc_x;
+    uint32_t edm_noc_y;
+    uint32_t termination_addr;
+};
+
+struct EdmLineFabricOpInterface {
+    enum Direction {
+        // Ascending chips in the sequence
+        FORWARD,
+
+        // Descending chips in the sequence
+        BACKWARD,
+    };
+
+    // Device ID -> EDM Builders
+    std::unordered_map<size_t, std::vector<FabricEriscDatamoverBuilder>> edm_builders_forward_direction;
+    std::unordered_map<size_t, std::vector<FabricEriscDatamoverBuilder>> edm_builders_backward_direction;
+
+    // Device ID -> link index
+    std::unordered_map<size_t, size_t> next_forward_direction_edm_available;
+    std::unordered_map<size_t, size_t> next_backward_direction_edm_available;
+
+    std::vector<Device*> device_sequence;
+    std::vector<Program*> programs;
+
+    size_t num_links;
```

Comment:
Lots of structs with uninitialized member variables. Please review this practice
https://github.com/tenstorrent/tt-metal/blob/main/best_practices.md#15-initialize-primitive-types-on-declaration